### PR TITLE
Merge supervisor

### PR DESCRIPTION
**Pull request by ReyhaneAskari**
This is the first for issue #5976. The supervisor class is removed and now the DestroyHandler should do the task.

### CODE REVIEW

**Comment by ReyhaneAskari** [comment]
I will do a rebase after the merge of faster_topo PR.

**Comment by ReyhaneAskari** [comment]
@nouiz , after the discussions with @lamblin last week, we found out that [here](https://github.com/Theano/Theano/blob/master/theano/compile/function_module.py#L185), the destroy handler should be given the protected variables. Since there were too many instantiations of DH we decided to have a minimal change to see if the tests would pass. So I created a protected attribute for fgraph and replaced the ```DH.protected``` with ```fgraph.protected```. We will see how the tests will pass.

**Comment by ReyhaneAskari** [comment]
jenkins test this.

**Comment by ReyhaneAskari** [comment]
Some of the tests were failing because it seems that the output of the graph was protected by the output guard. However, now we check the output of the graph inside the destroy handler as protected variables and we do not allow the inplace to happen on an output of the graph. So the output guard is not added as a node in the graph.

**Comment by nouiz** [comment]
I'm not sure setting protected as a FunctionGraph attribute/parameter is a good fix to the link you provided: https://github.com/Theano/Theano/blob/master/theano/compile/function_module.py#L185

I think we can just remove this loop, as now it is always done just bellow. (the Supervisor was converted to a DestroyHandler and is now always introduced).

File: theano/compile/function_module.py [hunk 1]
```diff
@@ -121,26 +121,6 @@ def fgraph_updated_vars(fgraph, expanded_inputs):
     return updated_vars
 
 
-class Supervisor:
```
Comment (ReyhaneAskari) [inline theano/compile/function_module.py]: Yes this is an option. I did see the pickle test failing and I was working on it yesterday. I thought maybe a cleaner option would be to unpickle the old test and remove the Supervisor from there and pickle it again. What do you think?

File: theano/compile/function_module.py [hunk 2]
```diff
@@ -121,26 +121,6 @@ def fgraph_updated_vars(fgraph, expanded_inputs):
...
-    Listener for FunctionGraph events which makes sure that no
-    operation overwrites the contents of protected Variables. The
-    outputs of the FunctionGraph are protected by default.
-
-    """
-
-    def __init__(self, protected):
-        self.protected = list(protected)
-
-    def validate(self, fgraph):
```
Comment (nouiz) [inline theano/compile/function_module.py]: New way of doing it:
- start over from the faster topo branch
- Modify destroyhandler to add a new fgraph.has_destroyers(list) (as fgraph.destroyers) that will return True is any of the variable in the list have a destroyers. (the old Superviser.validate check without the raise)
- Make that new fct do the Superviser.validate check when self.algo = cycle.
- Make sure all tests still pass
- Modify the new method to be faster (don't build the big data structure) when the algo is fast.

**Pull request closed without merging**